paddle/operators/sequence_concat_op.h

```diff
@@ -0,0 +1,167 @@
+#pragma once
+#include "paddle/framework/op_registry.h"
+#include "paddle/operators/strided_memcpy.h"
+
+namespace paddle {
+namespace operators {
+
+using Tensor = framework::Tensor;
+using LoDTensor = framework::LoDTensor;
+using LoD = framework::LoD;
+
+// Concat LoD, the initialized LoD of Output is lod(x0),
+// if axis is not 0, the LoD(Out) will be the same as Inputs, if axis is 0:
+// Case1:
+//  There is one level, the Output LoD will be modified:
+//  LoD(x0) = {{0,2,4}}
+//  LoD(x1) = {{0,1,5}}
+//  LoD(Out) = {{0,3,9}}
+// Case2:
+//  There is two level, and concat level is 1,
+//  the Output LoD will be modified as followed:
+//  LoD(x0) = {{0,2,4}, {0,1,2,3,4}}
+//  LoD(x1) = {{0,3,5}, {0,1,3,4,5}}
+//  LoD(Out) = {{0,5,9}, {0,1,2,4,5,6,7,8,9}}
+template <typename T>
+LoD concatLoD(const std::vector<const T*> ins, const size_t axis,
+              const size_t level) {
+  auto out_lod = ins[0]->lod();
+  const size_t n = ins.size();
+  if (axis == 0UL) {
+    if (level == 0) {
+      for (size_t i = 1; i < n; ++i) {
+        for (size_t j = 0; j < ins[i]->lod()[0].size(); ++j) {
+          out_lod[0][j] += ins[i]->lod()[0][j];
+        }
+      }
+    } else if (level == 1) {
+      PADDLE_ENFORCE_EQ(ins[0]->NumLevels(), 2UL,
+                        "If the level is 1, all of the inputs "
+                        "should be the the nested sequence.");
+      for (size_t i = 1; i < n; ++i) {
+        for (size_t j = 0; j < ins[i]->lod()[0].size(); ++j) {
+          out_lod[0].push_back(ins[i]->lod()[0][j]);
+        }
+        for (size_t j = 0; j < ins[i]->lod()[1].size(); ++j) {
+          out_lod[1][j] += ins[i]->lod()[1][j];
+        }
+      }
+    }
+  }
+  return out_lod;
+}
+
+template <typename Place, typename T>
+class SequenceConcatOpKernel : public framework::OpKernel<T> {
+ public:
+  void Compute(const framework::ExecutionContext& ctx) const override {
+    auto ins = ctx.MultiInput<LoDTensor>("X");
+    auto* out = ctx.Output<LoDTensor>("Out");
+    const size_t axis = static_cast<size_t>(ctx.Attr<int>("axis"));
+    const size_t level = static_cast<size_t>(ctx.Attr<int>("level"));
+    const size_t n = ins.size();
+
+    for (size_t i = 1; i < n; ++i) {
+      PADDLE_ENFORCE_EQ(ins[0]->NumLevels(), ins[i]->NumLevels(),
+                        "The level number of all the input LoDTensors "
+                        "should be the same.");
+      PADDLE_ENFORCE_EQ(ins[0]->dims().size(), ins[i]->dims().size(),
+                        "The dimensions size of all the input LoDTensors "
+                        "should be the same.");
+
+      const size_t dims_size = ins[i]->dims().size();
+      for (size_t j = 0; j < dims_size; ++j) {
+        if (j == axis) continue;
+        PADDLE_ENFORCE_EQ(ins[0]->dims()[j], ins[i]->dims()[j],
+                          "The dimensions of all the input LoDTensors "
+                          "except for the specify axis should be "
+                          "matched exactly.");
```

Comment:
Except for the dimension of the specified axis along which all the inputs are concatenated, dimensions of all the other axises of the input LoDTensors should be the same.

Reply:
Done.